tests/test_main.py

```diff
@@ -64,3 +65,64 @@ def test_get_credits_balance(self, requests_mock):
             json={"data": balance},
         )
         assert main.get_credits_balance() == balance
+
+
+class TestSession:
+    @pytest.fixture(scope="function")
+    def dummy_instance(self):
+        class DummyInstance:
+            session = main.Session()
+
+        return DummyInstance()
+
+    def test_should_fail_if_not_authenticated(self, dummy_instance):
+        with pytest.raises(main.UserNotAuthenticated):
+            _ = dummy_instance.session
+
+    def test_should_provide_session_if_authenticated(self, requests_mock, dummy_instance):
+        requests_mock.post("https://api.up42.com/oauth/token", json={"access_token": constants.TOKEN})
+        requests_mock.get(
+            url="https://api.up42.com/users/me",
+            json={"data": {"id": constants.WORKSPACE_ID}},
+        )
+        main.workspace.authenticate(username=constants.USER_EMAIL, password=constants.PASSWORD)
+        assert dummy_instance.session == main.workspace.auth.session
+
+
+class TestWorkspaceId:
+    @pytest.fixture
+    def dummy_instance_without_local_id(self, requests_mock):
+        requests_mock.post("https://api.up42.com/oauth/token", json={"access_token": constants.TOKEN})
+        requests_mock.get(
+            url="https://api.up42.com/users/me",
+            json={"data": {"id": constants.WORKSPACE_ID}},
+        )
+        main.workspace.authenticate(username=constants.USER_EMAIL, password=constants.PASSWORD)
+
+        class DummyInstance:
+            workspace_id = main.WorkspaceId()
+
+        yield DummyInstance()
+        main.workspace = main._Workspace()  # pylint: disable=protected-access
+
+    def test_should_provided_local_if_set_workspace(self):
+        class DummyInstance:
+            workspace_id = main.WorkspaceId()
+
+            def __init__(self, workspace_id=None):
+                if workspace_id:
+                    self.__dict__["workspace_id"] = workspace_id
```

Comment:
self.workspace_id = workspace_id

Reply:
And the class can be used in all cases

Reply:
And I'd recommend to just use a data class with eq=True and check whether it affects equality check